source-dynamodb/capture.go

```diff
@@ -0,0 +1,473 @@
+		// GetRecords request to an open shard may return no data even if we aren't at the end of
+		// it, and there is no guarantee of how many of these "no data" responses are needed to get
+		// to the next section of the shard with data. Reading only closed shards allows for knowing
+		// when we are at the end of the shard via the next shard iterator being `nil`.
+		//
+		// The horizon parameter is used to mitigate a possible race condition that can only exist
+		// if this capture were split into multiple Flow task shards (not currently supported, see
+		// TODO comment further down in this function). The challenge is that there is no way to
+		// tell the key range covered by individual backfill segments or stream shards. This means
+		// that one Flow task shard might end up processing keys for backfill segments that overlap
+		// with keys from stream shards for a different Flow task shard.
+		//
+		// There is a potential data race where an older backfill record might clobber a newer
+		// stream record:
+		// 1) Flow task shard A reads value of key `foo` from the table while backfilling.
+		// 2) A change to key `foo` in the table occurs. Task shard A will never read this key in
+		// the table again during its backfill.
+		// 3) Flow task shard B reads the change record of key `foo` from the stream.
+		// 4) Flow task shard B commits the change record to Flow.
+		// 5) Flow task shard A commits the now stale value it read to Flow.
+		//
+		// When running these "catch up" reads, stream change records newer than the horizon will be
+		// held back to prevent this possible data race. The assumption is that the Flow task shard
+		// responsible for backfilling the table will not be delayed in actually committing the
+		// values read to Flow for longer than the specified horizon.
+		log.WithFields(log.Fields{
+			"table": t.tableName,
+		}).Info("catching up streams for table")
+		if err := c.catchupStreams(ctx, t, backfillStreamingHorizon); err != nil {
+			return fmt.Errorf("catching up streams for table '%s': %w", t.tableName, err)
+		}
+
+		// Alternate catching up streams with backfilling from the table. The backfill runs for the
+		// duration provided by backfillDuration. Alternating between backfilling a streaming is a
+		// little simpler than backfilling continuously concurrently with streaming because of the
+		// need to manage the horizon time for streaming while the backfill is in progress.
+		log.WithFields(log.Fields{
+			"table":            t.tableName,
+			"segments":         t.totalBackfillSegments,
+			"rcus":             t.effectiveRcus,
+			"backfillDuration": backfillDuration.String(),
+		}).Info("backfilling table")
+		if err := c.backfill(ctx, t, backfillDuration); err != nil {
+			return fmt.Errorf("backfilling table '%s': %w", t.tableName, err)
+		}
+	}
+
+	// TODO(whb): If we were ever going to allow these Flow consumer shards to be split, we would
```

Comment:
This is all fine / no changes required, but something to ponder:

Does dynamodb offer a monotonic version or time attached to records from both streaming and backfill? If so, we could achieve eventual consistency by using the `max` reduction annotation while also concurrently doing backfills and streaming.

Reply:
Unfortunately it doesn't have any kind of version or timestamp for backfill records, and all you get from stream records is an "approximate" timestamp for the record. The closest thing we could get would be the use the approximate stream record timestamp and a wall clock time for when the backfill scan was executed, which I think would only work _most_ of the time.

Reply:
Blarg. Okay, sounds good then considering.